File: project/block_manager/services/nodes/pytorch/conv1d.py
"""PyTorch Conv1D Node Definition"""

from typing import Dict, List, Optional, Any
from ..base import NodeDefinition, NodeMetadata, ConfigField, TensorShape, Framework


class Conv1DNode(NodeDefinition):
    """1D Convolution layer"""
    
    @property
    def metadata(self) -> NodeMetadata:
        return NodeMetadata(
            type="conv1d",
            label="Conv1D",
            category="advanced",
            color="var(--color-purple)",
            icon="WaveSquare",
            description="1D convolutional layer",
            framework=Framework.PYTORCH
        )
    
    @property
    def config_schema(self) -> List[ConfigField]:
        return [
            ConfigField(
                name="out_channels",
                label="Output Channels",
                type="number",
                required=True,
                min=1,
                description="Number of output channels"
            ),
            ConfigField(
                name="kernel_size",
                label="Kernel Size",
                type="number",
                default=3,
                min=1,
                description="Size of the convolving kernel"
            ),
            ConfigField(
                name="stride",
                label="Stride",
                type="number",
                default=1,
                min=1,
                description="Stride of the convolution"
            ),
            ConfigField(
                name="padding",
                label="Padding",
                type="number",
                default=0,
                min=0,
                description="Zero padding on both sides"
            ),
            ConfigField(
                name="dilation",
                label="Dilation",
                type="number",
                default=1,
                min=1,
                description="Spacing between kernel elements"
            ),
            ConfigField(
                name="bias",
                label="Use Bias",
                type="boolean",
                default=True,
                description="Add learnable bias"
            )
        ]
    
    def compute_output_shape(
        self,
        input_shape: Optional[TensorShape],
        config: Dict[str, Any]
    ) -> Optional[TensorShape]:
        if not input_shape or len(input_shape.dims) != 3:
            return None
        
        batch, in_channels, length = input_shape.dims
        out_channels = int(config.get("out_channels", in_channels))
        kernel_size = int(config.get("kernel_size", 3))
        stride = int(config.get("stride", 1))
        padding = int(config.get("padding", 0))
        dilation = int(config.get("dilation", 1))
        
        # Calculate output length
        out_length = ((length + 2 * padding - dilation * (kernel_size - 1) - 1) // stride) + 1
        
        return TensorShape(
            dims=[batch, out_channels, out_length],
            description=f"Conv1D({out_channels})"
        )
    
    def validate_incoming_connection(
        self,
        source_node_type: str,
        source_output_shape: Optional[TensorShape],
        target_config: Dict[str, Any]
    ) -> Optional[str]:
        # Allow connections from input/dataloader without shape validation
        if source_node_type in ("input", "dataloader"):
            return None
        
        # Empty and custom nodes are flexible
        if source_node_type in ("empty", "custom"):
            return None
        
        # Validate 3D input (N, C, L)
        return self.validate_dimensions(
            source_output_shape,
            3,
            "[batch, channels, length]"
        )

File: project/frontend/src/lib/nodes/definitions/pytorch/conv1d.ts
/**
 * PyTorch Conv1D Layer Node Definition
 */

import { NodeDefinition } from '../../base'
import { NodeMetadata, BackendFramework } from '../../contracts'
import { TensorShape, BlockConfig, ConfigField, BlockType } from '../../../types'

export class Conv1DNode extends NodeDefinition {
  readonly metadata: NodeMetadata = {
    type: 'conv1d',
    label: 'Conv1D',
    category: 'advanced',
    color: 'var(--color-purple)',
    icon: 'WaveSquare',
    description: '1D convolutional layer',
    framework: BackendFramework.PyTorch
  }

  readonly configSchema: ConfigField[] = [
    {
      name: 'out_channels',
      label: 'Output Channels',
      type: 'number',
      required: true,
      min: 1,
      description: 'Number of output channels'
    },
    {
      name: 'kernel_size',
      label: 'Kernel Size',
      type: 'number',
      default: 3,
      min: 1,
      description: 'Size of the convolving kernel'
    },
    {
      name: 'stride',
      label: 'Stride',
      type: 'number',
      default: 1,
      min: 1,
      description: 'Stride of the convolution'
    },
    {
      name: 'padding',
      label: 'Padding',
      type: 'number',
      default: 0,
      min: 0,
      description: 'Zero padding on both sides'
    },
    {
      name: 'dilation',
      label: 'Dilation',
      type: 'number',
      default: 1,
      min: 1,
      description: 'Spacing between kernel elements'
    },
    {
      name: 'bias',
      label: 'Use Bias',
      type: 'boolean',
      default: true,
      description: 'Add learnable bias'
    }
  ]

  computeOutputShape(inputShape: TensorShape | undefined, config: BlockConfig): TensorShape | undefined {
    if (!inputShape || !config.out_channels) {
      return undefined
    }

    if (inputShape.dims.length !== 3) {
      return undefined
    }

    const [batch, _, length] = inputShape.dims as number[]

    const kernel = (config.kernel_size ?? 3) as number
    const stride = (config.stride ?? 1) as number
    const padding = (config.padding ?? 0) as number
    const dilation = (config.dilation ?? 1) as number

    const outLength = Math.floor((length + 2 * padding - dilation * (kernel - 1) - 1) / stride) + 1

    return {
      dims: [batch, config.out_channels as number, outLength],
      description: `Conv1D(${config.out_channels})`
    }
  }

  validateIncomingConnection(
    sourceNodeType: BlockType,
    sourceOutputShape: TensorShape | undefined,
    targetConfig: BlockConfig
  ): string | undefined {
    if (sourceNodeType === 'input' || sourceNodeType === 'dataloader') {
      return undefined
    }

    if (sourceNodeType === 'empty' || sourceNodeType === 'custom') {
      return undefined
    }

    return this.validateDimensions(sourceOutputShape, {
      dims: 3,
      description: '[batch, channels, length]'
    })
  }
}
